test/e2e/e2e_suite_test.go

```diff
@@ -403,10 +408,102 @@ func tearDown(bootstrapClusterProvider bootstrap.ClusterProvider, bootstrapClust
 	}
 }
 
+// resolveKubernetesVersions looks at Kubernetes versions set as variables in the e2e config and sets them to a valid k8s version
+// that has an existing capi offer image available. For example, if the version is "stable-1.22", the function will set it to the latest 1.22 version that has a published reference image.
+func resolveKubernetesVersions(config *clusterctl.E2EConfig) {
+	skus := getAllCAPIImageSkus(context.TODO(), os.Getenv(AzureLocation))
+	versions := parseImageSkuNames(skus)
+
+	if config.HasVariable(capi_e2e.KubernetesVersion) {
+		resolveKubernetesVersion(config, versions, capi_e2e.KubernetesVersion)
+	}
+	if config.HasVariable(capi_e2e.KubernetesVersionUpgradeFrom) {
+		resolveKubernetesVersion(config, versions, capi_e2e.KubernetesVersionUpgradeFrom)
+	}
+	if config.HasVariable(capi_e2e.KubernetesVersionUpgradeTo) {
+		resolveKubernetesVersion(config, versions, capi_e2e.KubernetesVersionUpgradeTo)
+	}
+}
+
+func resolveKubernetesVersion(config *clusterctl.E2EConfig, versions semver.Versions, varName string) {
+	v := getLatestSkuForMinor(config.GetVariable(varName), versions)
+	if _, ok := os.LookupEnv(varName); ok {
+		Expect(os.Setenv(varName, v)).To(Succeed())
```

Comment:
At first I didn't have this (only the config.Variables part below), but it didn't work. Unfortunately that's the only way to modify the config since the capi framework gives precedence to env variables if they are set: https://github.com/kubernetes-sigs/cluster-api/blob/main/test/framework/clusterctl/e2e_config.go#L553

FWIW, this is similar to what CAPI does for CAPD e2e tests https://github.com/kubernetes-sigs/cluster-api/blob/main/scripts/ci-e2e-lib.sh#L56

Reply:
oh it's a capi constraint. Thanks for the explanation!